Backpack/src/main/java/net/skyscanner/backpack/button/internal/BpkButtonBase.kt

```diff
@@ -0,0 +1,263 @@
+package net.skyscanner.backpack.button.internal
+
+import android.content.Context
+import android.graphics.Bitmap
+import android.graphics.Canvas
+import android.graphics.drawable.BitmapDrawable
+import android.graphics.drawable.Drawable
+import android.graphics.drawable.VectorDrawable
+import android.util.AttributeSet
+import android.util.Log
+import android.view.Gravity
+import android.widget.TextView
+import androidx.annotation.Dimension
+import androidx.annotation.DrawableRes
+import androidx.appcompat.content.res.AppCompatResources
+import androidx.appcompat.widget.AppCompatButton
+import androidx.core.content.ContextCompat
+import androidx.core.graphics.drawable.DrawableCompat
+import androidx.vectordrawable.graphics.drawable.VectorDrawableCompat
+import net.skyscanner.backpack.R
+import net.skyscanner.backpack.text.BpkText
+import net.skyscanner.backpack.util.darken
+import net.skyscanner.backpack.util.wrapContextWithDefaults
+
+internal const val ICON_POSITION_START = 0
+internal const val ICON_POSITION_END = 1
+internal const val ICON_POSITION_ICON_ONLY = 2
+
+internal const val INVALID_RES = -1
+
+open abstract class BpkButtonBase internal constructor(
+  context: Context,
+  attrs: AttributeSet?,
+  defStyleAttr: Int
+) : AppCompatButton(wrapContextWithDefaults(context), attrs, defStyleAttr) {
+
+  private var isInitialized = false
+
+  private val textMeasurement: TextMeasurement by lazy { TextMeasurement(paint) }
+  internal val tokens = ButtonTokens(this.context)
+
+  private val font = BpkText.getFont(this.context, BpkText.SM, BpkText.Weight.EMPHASIZED)
+
+  private var overrideTextColor: Int? = null
+
+  private var _buttonTextColor: Int = ContextCompat.getColor(this.context, R.color.bpkWhite)
+  protected var buttonTextColor: Int
+    get() = overrideTextColor ?: _buttonTextColor
+    set(value) {
+      _buttonTextColor = value
+      updateSelf()
+    }
+
+  @Dimension
+  private var originalStartPadding: Int = 0
+
+  @Dimension
+  private var originalEndPadding: Int = 0
+
+  private var _iconPosition: Int = ICON_POSITION_END
+  open var iconPosition: Int
+    get() = _iconPosition
+    set(value) {
+      _iconPosition = value
+      updateSelf()
+    }
+
+  private var _icon: Drawable? = null
+  var icon: Drawable? = null
+    get() = _icon
+    set(value) {
+      if (value != field) {
+        field = value
+
+        value?.let {
+          _icon = convertAndResize(it)
```

Comment:
That's a bad practice to call setters recursively. Why would not use field?

Reply:
I know that this code was in the prev codebase, but that's the original reason we convert a drawable to a bitmap here?

Reply:
We convert it to set it to the proper size, our icons are 24pt by default and they need to be 16pt for the button.

Reply:
Sorry, it's not recursive, my mistake. We can use LayerDrawable instead of Bitmap conversion for that.

Reply:
We can try again but as far as I remember it din't work. But fell free to create a card for that.